### PR TITLE
Fixes all the reported block disabling issues

### LINKED ISSUE #2954: Rainbow blocks still work after disabling them
## :round_pushpin: Description (REQUIRED)
When disabling the rainbow blocks they still work. Work as in change colors
This has been tested with all SF rainbow blocks

## :bookmark_tabs: Steps to reproduce the Issue (REQUIRED)
https://youtu.be/oryRBTfbh1o
In the video it shows the full steps to reproduce this issue

## :bulb: Expected behavior (REQUIRED)
to not change colors after toggling them off in console

## :scroll: Server Log
N/a

## :open_file_folder: /error-reports/ Folder
N/A

## :compass: Environment (REQUIRED)

image

### LINKED ISSUE #2957: Auto crafter still working after being disabled
## :round_pushpin: Description (REQUIRED)
So i was also testing the enhanced auto crafter. soo it works when disabled. what works
Set a recipe in the enhanced auto crafter
stop the server
disable the auto crafter
start the server
you can open the gui.
You can also toggle of and on stuff inside the gui. 
but when the the recipe is turned of the machine breaks after a certain amount of time.
but it does display the wrong message. when trying to use the disabled auto crafter.
for example right clicking with empty hand shows 
image
shift right clicking shows this message
image
but when you do it with an item that shouldnt work on it like a vanilla chest it shows this
image
Another update on this when the machine is disabled and you interact with it you can steal the items inside of it
i just obtained a barrier

## :bookmark_tabs: Steps to reproduce the Issue (REQUIRED)
Thanks to walshy for uploading the vid blobs.are-pretty.sexy/7FbFgdR.mp4

## :bulb: Expected behavior (REQUIRED)

To not work after being disabled

## :scroll: Server Log
N/A

## :open_file_folder: /error-reports/ Folder
N/A

## :compass: Environment (REQUIRED)
image

### LINKED ISSUE #2958: Carbonado edge furnace still working after being disabled
## :round_pushpin: Description (REQUIRED)
When the carbonado edged furnace has been disabled with items still inside of it. It will still work with the speed the machine is normally it wont work on default furnace speed as you would expect.

## :bookmark_tabs: Steps to reproduce the Issue (REQUIRED)
Place down carbonado edged furnace with items inside of it.
Stop server
disable the furnace
start server
Furnace is still smelting at the speed of carbonado edged but it shouldnt

## :bulb: Expected behavior (REQUIRED)
Furnace to act like a normal furnace or to just being disabled and canceling the smelting

## :scroll: Server Log
N/A

## :open_file_folder: /error-reports/ Folder
N/A

## :compass: Environment (REQUIRED)
image

### LINKED ISSUE #2959: Blockplacer still functions after being disabled
## :round_pushpin: Description (REQUIRED)
Blockplacer still functioning even when disabled if it has items inside of it

## :bookmark_tabs: Steps to reproduce the Issue (REQUIRED)
Place blockplacer
place items inside of it
stop server
disable the blockplacer
start server
switch the lever to power the blockplacer
Sees blocks being placed

## :bulb: Expected behavior (REQUIRED)
Dispenser to stop working fully or to just work as a dispenser

## :scroll: Server Log
N/A

## :open_file_folder: /error-reports/ Folder
N/A

## :compass: Environment (REQUIRED)
image

### LINKED ISSUE #2966: Elevator Plate is still working after being disabled
## :round_pushpin: Description (REQUIRED)

Elevator Plate is still working after being disabled

## :bookmark_tabs: Steps to reproduce the Issue (REQUIRED)

1. Place down Elevator Plates
2. Disable it on items.yml
3. It still works

https://user-images.githubusercontent.com/34732414/115138276-44b04880-a055-11eb-9ce4-016b79981510.mp4

## :bulb: Expected behavior (REQUIRED)

It shouldn't work.

## :scroll: Server Log

No relevant server log.

## :open_file_folder: /error-reports/ Folder

No error log.

## :compass: Environment (REQUIRED)

image

### PR DESCRIPTION
**Pull request by WalshyDev**
## Description
Added isDisabledIn in places where it was missing and causing unexpected behaviour. I'm sure there are more but these are all the reported ones.

The most up to date build with these fixes is here: https://cdn.discordapp.com/attachments/828939960769708032/833372572753330227/Slimefun_disabled-block-fixes.jar

## Changes
* Added isDisabledIn in AutoCrafterListener to prevent usage if disabled
* Added isDisabledIn in DispenserListener to prevent things like the Block Placer from functioning
* Adedd isDisabledIn in EnhancedFurnaceListener to prevent the speed changing if disabled
* Added isDisabledIn in TeleporterListener to prevent the teleporter and elevator plate
* Added isDisabledIn in TickerTask#tickLocation to prevent block ticking if disabled

## Related Issues
* Fixes #2957
* Fixes #2958
* Fixes #2954
* Fixes #2959
* Fixes #2957
* Fixes #2966

## Checklist

- [x] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [x] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [x] I followed the existing code standards and didn't mess up the formatting.
- [ ] I did my best to add documentation to any public classes or methods I added.
- [ ] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

**Comment by WalshyDev** [comment]
Thank you to @variananora for helping me test and finding issues

**Comment by J3fftw1** [comment]
> Thank you to @variananora for helping me test and finding issues

Pancries, i have reported a bunch aswell 😭

**Comment by WalshyDev** [comment]
> > Thank you to @variananora for helping me test and finding issues
> 
> Pancries, i have reported a bunch aswell

I more meant finding issues in the test builds but sure, I'll humor you, thank you for finding some of the original issues Jeff *head pats*

**Comment by J3fftw1** [comment]
> > > Thank you to @variananora for helping me test and finding issues
> > 
> > 
> > Pancries, i have reported a bunch aswell
> 
> I more meant finding issues in the test builds but sure, I'll humor you, thank you for finding some of the original issues Jeff _head pats_

https://media.giphy.com/media/5tmRHwTlHAA9WkVxTU/giphy.gif
🤗

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/AutoCrafterListener.java [hunk 1]
```diff
@@ -50,7 +50,8 @@ public void onInteract(PlayerRightClickEvent e) {
 
             SlimefunItem block = slimefunBlock.get();
 
-            if (block instanceof AbstractAutoCrafter) {
+            // Fixes #2957
+            if (block instanceof AbstractAutoCrafter && !block.isDisabledIn(e.getPlayer().getWorld())) {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/AutoCrafterListener.java:54]: This simple check won't send a message to let the player know what's wrong.
On that occasion, we don't check for unlocked researches either, so this should rather be `.canUse(...)`

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/TeleporterListener.java [hunk 2]
```diff
@@ -34,31 +34,32 @@ public void onPressurePlateEnter(PlayerInteractEvent e) {
...
     }
 
     @ParametersAreNonnullByDefault
-    private boolean isTeleporterPad(String id, Block b, UUID uuid) {
-        if (id.equals(SlimefunItems.GPS_ACTIVATION_DEVICE_SHARED.getItemId())) {
+    private boolean isTeleporterPad(SlimefunItem item, Block b, UUID uuid) {
+        if (item.getId().equals(SlimefunItems.GPS_ACTIVATION_DEVICE_SHARED.getItemId())) {
             return true;
-        } else if (id.equals(SlimefunItems.GPS_ACTIVATION_DEVICE_PERSONAL.getItemId())) {
+        } else if (item.getId().equals(SlimefunItems.GPS_ACTIVATION_DEVICE_PERSONAL.getItemId())) {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/TeleporterListener.java:62]: This hardcoded id check is really not optimal.

Is there a Teleporter Pad class? If not we should probably add that and a `hasAccess(Player, Block)` method then to allow modifications.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerTask.java [hunk 3]
```diff
@@ -99,14 +99,7 @@ public void run() {
...
             // Fixes #2576 - Remove any deleted instances of BlockStorage
-            Iterator<BlockStorage> worlds = SlimefunPlugin.getRegistry().getWorlds().values().iterator();
-            while (worlds.hasNext()) {
-                BlockStorage storage = worlds.next();
-
-                if (storage.isMarkedForRemoval()) {
-                    worlds.remove();
-                }
-            }
+            SlimefunPlugin.getRegistry().getWorlds().values().removeIf(BlockStorage::isMarkedForRemoval);
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerTask.java:102]: What is the performance on this?

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerTask.java [hunk 4]
```diff
@@ -154,13 +147,14 @@ private void tickLocation(@Nonnull Set<BlockTicker> tickers, @Nonnull Location l
         Config data = BlockStorage.getLocationInfo(l);
         SlimefunItem item = SlimefunItem.getByID(data.getString("id"));
 
-        if (item != null && item.getBlockTicker() != null) {
+        // Fixes #2954
+        if (item != null && !item.isDisabledIn(l.getWorld()) && item.getBlockTicker() != null) {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerTask.java:151]: Checking this on every tick is extremely bad and can weigh everything down.
We should rather not even mark this as a ticker if it is disabled anyway, so it won't even show up in this `Set`.

File: CHANGELOG.md [hunk 5]
```diff
@@ -22,6 +23,15 @@
...
 - [Release Candidate 1 (26 Sep 2019)](#release-candidate-1-26-sep-2019)
 
+## Release Candidate 23 (TBD)
+
+#### Additions
+
+#### Changes
+
+#### Fixes
+* Fixes numerous blocks still functioning if disabled
```
Comment (TheBusyBiscuit) [inline CHANGELOG.md:33]: Please list the actual issue numbers here

File: CHANGELOG.md [hunk 6]
```diff
@@ -22,6 +23,15 @@
 - [Release Candidate 2 (29 Sep 2019)](#release-candidate-2-29-sep-2019)
 - [Release Candidate 1 (26 Sep 2019)](#release-candidate-1-26-sep-2019)
 
+## Release Candidate 23 (TBD)
+
+#### Additions
+
```
Comment (TheBusyBiscuit) [inline CHANGELOG.md:29]: Can also add #2965 here while we are at it :eyes:

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerTask.java (hunk 4 above)
Comment (variananora) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerTask.java:151]: Tested with 101x101x101 Rainbow Glass:

Rainbow Glass is enabled:
image

Rainbow Glass is disabled:
image

idk if this is relevant or not

Comment (TheBusyBiscuit) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerTask.java:151]: The impact might be small, but it is avoidable and any optimizations are good, small and large.

File: CHANGELOG.md (hunk 6 above)
Comment (TheBusyBiscuit) [reply CHANGELOG.md:29]: Nevermind this one, added it myself in a commit now

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/TeleporterListener.java (hunk 2 above)
Comment (TheBusyBiscuit) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/TeleporterListener.java:62]: Refactored in https://github.com/Slimefun/Slimefun4/commit/a3a9ee4b909a98696e96d1bd4c5551204986ec18

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/AutoCrafterListener.java (hunk 1 above)
Comment (WalshyDev) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/AutoCrafterListener.java:54]: canUse doesn't check world, should I modify it to do so? That would probably fix even more disabled things working (if disabled for a certain world)

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerTask.java (hunk 3 above)
Comment (WalshyDev) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerTask.java:102]: they do the exact same code. The only difference is removeIf obviously takes a Predicate so runs that. This way is just much cleaner

https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/util/Collection.java#L571-L582

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerTask.java (hunk 4 above)
Comment (WalshyDev) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerTask.java:151]: Will do this but also worth noting that this wont cover different worlds. Can't think of a better way to check it than every tick in this case.

**Comment by WalshyDev** [comment]
oh right, I can't use isDisabled in addItemHandler since it isn't registered yet haha

is there a better way around this?

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/AutoCrafterListener.java (hunk 1 above)
Comment (TheBusyBiscuit) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/AutoCrafterListener.java]: It does...
https://github.com/Slimefun/Slimefun4/blob/master/src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/SlimefunItem.java#L1084-L1090

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerTask.java (hunk 4 above)
Comment (TheBusyBiscuit) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerTask.java]: Tickers are handled per-world anyway.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/AutoCrafterListener.java [hunk 7]
```diff
@@ -50,7 +50,8 @@ public void onInteract(PlayerRightClickEvent e) {
 
             SlimefunItem block = slimefunBlock.get();
 
-            if (block instanceof AbstractAutoCrafter) {
+            // Fixes #2957
+            if (block instanceof AbstractAutoCrafter && !block.canUse(e.getPlayer(), true)) {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/AutoCrafterListener.java:54]: Why is this inverted?

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]

**Comment by WalshyDev** [comment]
> 👀

you never answered me! PepeHands

> oh right, I can't use isDisabled in addItemHandler since it isn't registered yet haha
> is there a better way around this?

The issue is that addItemHandler will [register the ticker right away](https://github.com/Slimefun/Slimefun4/blob/27224c55eb69f058d8dc0b0321a0628adba2727f/src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/SlimefunItem.java#L821) rather than waiting until SlimefunItem#onEnable. I'm not sure why it's treated differently so I don't really want to change this behaviour as of now, do you have a good workaround for this? Should I just change it and get people to test the code?

**Comment by TheBusyBiscuit** [comment]
Alright, I fixed up some things in here.

**Review by svr333 — APPROVED** [review]
Looks good
Though are the `// fixes <issue>` comments necessary everywhere?
I feel like a few days after this PR was merged they'll just be redundant

**Comment by TheBusyBiscuit** [comment]
> Looks good

@svr333 
You do realize this has hundreds of test failures and fails to compile, right? :LUL:

**Comment by svr333** [comment]
yes I saw those but I thought those were from a merge issue or something

**Comment by TheBusyBiscuit** [comment]
Closing this as "stale" after over a month of inactivity.

**Pull request closed without merging**